tensorstore/kvstore/s3/s3_key_value_store.cc

```diff
@@ -0,0 +1,1264 @@
+        read_result.state = kvstore::ReadResult::kMissing;
+        return read_result;
+      case 412:
+        // "Failed precondition": indicates the ifGenerationMatch condition
+        // did not hold.
+        // NOTE: This is returned even when the object does not exist.
+        read_result.stamp.generation = StorageGeneration::Unknown();
+        return read_result;
+      case 304:
+        // "Not modified": indicates that the ifGenerationNotMatch condition
+        // did not hold.
+        read_result.stamp.generation = options.if_not_equal;
+        return read_result;
+    }
+
+    size_t payload_size = httpresponse.payload.size();
+    if (httpresponse.status_code != 206) {
+      // This may or may not have been a range request; attempt to validate.
+      TENSORSTORE_ASSIGN_OR_RETURN(auto byte_range,
+                                   options.byte_range.Validate(payload_size));
+      read_result.state = kvstore::ReadResult::kValue;
+      read_result.value =
+          internal::GetSubCord(httpresponse.payload, byte_range);
+    } else {
+      // Server should return a parseable content-range header.
+      TENSORSTORE_ASSIGN_OR_RETURN(auto content_range_tuple,
+                                   ParseContentRangeHeader(httpresponse));
+
+      if (options.byte_range.inclusive_min !=
+              std::get<0>(content_range_tuple) ||
+          payload_size != options.byte_range.size().value_or(payload_size)) {
+        // Return an error when the response does not start at the requested
+        // offset of when the response is smaller than the desired size.
+        return absl::OutOfRangeError(tensorstore::StrCat(
+            "Requested byte range ", options.byte_range,
+            " was not satisfied by s3 response of size ", payload_size));
+      }
+      assert(payload_size == std::get<2>(content_range_tuple));
+      read_result.state = kvstore::ReadResult::kValue;
+      read_result.value = httpresponse.payload;
+    }
+
+    // TODO: Avoid parsing the entire metadata & only extract the
+    // generation field.
+    ObjectMetadata metadata;
+    SetObjectMetadataFromHeaders(httpresponse.headers, &metadata);
+
+    // read_result.stamp.generation =
+    //     StorageGeneration::FromUint64(metadata.generation);
+    return read_result;
+  }
+};
+
+Future<kvstore::ReadResult> S3KeyValueStore::Read(Key key,
+                                                   ReadOptions options) {
+  s3_read.Increment();
+  if (!IsValidObjectName(key)) {
+    return absl::InvalidArgumentError("Invalid S3 object name");
+  }
+    // TODO(sjperkins). Introduce S3 versioning logic here
+  // if (!IsValidStorageGeneration(options.if_equal) ||
+  //     !IsValidStorageGeneration(options.if_not_equal)) {
+  //   return absl::InvalidArgumentError("Malformed StorageGeneration");
+  // }
+
+  auto encoded_object_name = UriObjectKeyEncode(key);
+  std::string resource = tensorstore::internal::JoinPath(endpoint_, "/",
```

Comment:
Just use StrCat, not JoinPath, for urls.  If `encoded_object_name` starts with a "/" that needs to be preserved.

Reply:
Done